doc/advanced/content/executor_design.rst

```diff
@@ -0,0 +1,417 @@
+
+This was was one of the initial design considerations. It seemed like an attractive choice due to simplicity,
+could choose the allowed overload resolution to choose the best option at compile time by allowing
+tags to be inherited. These tags would serve as placeholders till executors were standardized.
+They lacked all the features of executors and offered no customization. Once executors were even standardized
+large parts of the codebase would need to be refactored to support executors.
+
+* Base Executor
+
+The idea was to have single base executor from which all executors would derive from. This would allow
+all the common cod to be shared among all the executors. The base executors has a CRTP based design which
+is used to access the properties of the derived executor. It also allowed simplification in many areas
+such as being able to reference any derived executor. Basic properties of all executors like
+copy constructors, overloaded equality operators.
+The CRTP mechanism had some restrictions in the sense that there was still a need for templates and
+their were concerns regarding explicitly passing all the properties to the base class using CRTP as
+it was felt to be unnecessary. Having inheritance also introduces runtime polymorphism which was against
+one of the design consideration i.e. have everything compile time and avoid any overhead.
+The base executor besides a few common functionality didn't do much and was discarded.
+
+* Overload Mechanism
+
+In this design, the supported executor properties specified as templated parameters were inherited from.
+With this design executors would not need to provide the `require` or `prefer` and `query` member functions
+inside the executor for each property it wants to allows association with and support customizing.
+This design was not fully explore, so the potential drawbacks are not know completely. A snippet
+is available `here <https://godbolt.org/z/zhKM6e>`_. It was decided to not go with this design forward
+due to the potential complications, minimal support of properties for executors in PCL and deviation
+from the property customization mechanism design implemented in other implementations of the executor design
+proposal mentioned above in the document. Once more properties to be used w.r.t PCL are made available and
+use cases for these properties comes up. This design can be considered and looked at.
+and
+
+* Unified Shape with Compile Time Support
+Since each executor could have a different shape type, a unified shape type would provide a more uniform
+API. The compile time support for shapes would also bring performance benefits. The compile time aspect
+was ditched since in most scenarios the value of the shape was determined at runtime and rarely at compile
+time. The unified shape was also deemed unnecessary since for most executors `std::size_t` or and `int` was
+sufficient and only a custom shape was needed for GPU (CUDA) code for which tje shape could explicitly
```

Comment:
a custom shape was needed only for GPU

for which the shape could be explicitly